common/domain/handler.go

```diff
@@ -815,242 +895,32 @@ func (d *handlerImpl) FailoverDomain(
-			)
-
-			err = updateFailoverHistoryInDomainData(info, d.config, NewFailoverEvent(
-				now,
-				failoverType,
-				&currentActiveCluster,
-				updateRequest.ActiveClusterName,
-				nil,
-				nil,
-			))
-			if err != nil {
-				d.logger.Warn("failed to update failover history", tag.Error(err))
-			}
-		}
-
-		// case 2. active-passive domain is being migrated to active-active
-		if !wasActiveActive && isActiveActive {
-			// for active-passive to active-active migration,
-			// we increment failover version so top level failoverVersion is updated and domain data is replicated.
-
-			failoverVersion = d.clusterMetadata.GetNextFailoverVersion(
-				replicationConfig.ActiveClusterName,
-				failoverVersion+1, //todo: (active-active): Let's review if we need to increment
-				// this for cluster-attr failover changes. It may not be necessary to increment
-				updateRequest.Name,
-			)
-
-			d.logger.Debug("active-passive domain is being migrated to active-active",
-				tag.WorkflowDomainName(info.Name),
-				tag.Dynamic("failover-version", failoverVersion),
-				tag.Dynamic("failover-type", failoverType),
-			)
-
-			err = updateFailoverHistoryInDomainData(info, d.config, NewFailoverEvent(
-				now,
-				failoverType,
-				&currentActiveCluster,
-				updateRequest.ActiveClusterName,
-				nil,
-				replicationConfig.ActiveClusters,
-			))
-			if err != nil {
-				d.logger.Warn("failed to update failover history", tag.Error(err))
-			}
-		}
-
-		// case 3. active-active domain's ActiveClusters is changed
-		if wasActiveActive && isActiveActive {
-			// top level failover version is not used for task versions for active-active domains but we still increment it
-			// to indicate there was a change in replication config
-			failoverVersion = d.clusterMetadata.GetNextFailoverVersion(
-				replicationConfig.ActiveClusterName,
-				failoverVersion+1, //todo: (active-active): Let's review if we need to increment
-				// this for cluster-attr failover changes. It may not be necessary to increment
-				updateRequest.Name,
-			)
-
-			d.logger.Debug("active-active domain failover",
-				tag.WorkflowDomainName(info.Name),
-				tag.Dynamic("failover-version", failoverVersion),
-				tag.Dynamic("failover-type", failoverType),
-			)
-
-			err = updateFailoverHistoryInDomainData(info, d.config, NewFailoverEvent(
-				now,
-				failoverType,
-				&currentActiveCluster,
-				nil,
-				currentActiveClusters,
-				replicationConfig.ActiveClusters,
-			))
-			if err != nil {
-				d.logger.Warn("failed to update failover history", tag.Error(err))
-			}
-		}
-
-		failoverNotificationVersion = notificationVersion
-	}
-
-	lastUpdatedTime = now
-
-	updateReq := createUpdateRequest(
-		info,
-		config,
-		replicationConfig,
-		configVersion,
-		failoverVersion,
-		failoverNotificationVersion,
-		gracefulFailoverEndTime,
-		previousFailoverVersion,
-		lastUpdatedTime,
+	response, err := d.handleFailoverRequest(
+		ctx,
+		failoverRequest.ToUpdateDomainRequest(),
```

Comment:
chose to map to the UpdateDomain request struct here rather than the reverse because the `FailoverDomain` API is actually a subset of the UpdateDomain failover capabilities, we didn't include any graceful failover fields for it, so it has to be this way